### PR TITLE
Describe api for sequential test steps

### PR DESCRIPTION
**Pull request by christophsturm**

### CODE REVIEW

File: failgood/jvm/src/failgood/ContextDSL.kt [hunk 1]
```diff
@@ -110,4 +112,6 @@ interface ContextDSL<GivenType> : ResourcesDSL {
      */
     suspend fun test(name: String, tags: Set<String> = setOf(), function: TestLambda<GivenType>) =
         it(name, tags, function)
+
+    suspend fun describeSteps(name: String, function: suspend It<Unit>.() -> Unit) {}
```
Comment (mervyn-mccreight) [inline failgood/jvm/src/failgood/ContextDSL.kt:116]: Accidental empty default implementation? 🤔

**Comment by mervyn-mccreight** [comment]
One general thought about this is - we open the doors for un-isolated tests here - wasn't a design goal of failgood to always strive for isolated and parallel tests? 🤔

File: failgood/jvm/src/failgood/ContextDSL.kt (hunk 1 above)
Comment (christophsturm) [reply failgood/jvm/src/failgood/ContextDSL.kt:116]: just to flesh out the api

**Comment by christophsturm** [comment]
@mervyn-mccreight i hope the tests now explain better how its supposed to work. also `describeSteps` is just a temp name.

**Comment by mervyn-mccreight** [comment]
Would it make sense to have a separate DSL for unit and integration testing? I think these are two separate use-cases 🤔

**Comment by christophsturm** [comment]
for now I'm just wondering how the api could look like, not sure when I'm going to implement it. But it sure would be useful to split up our e2e tests into separate steps

**Comment by mervyn-mccreight** [comment]
I'm curious of this is just for e2e tests if this really should be in scope of FailGood, since e2e tests are a very special case. FailGood as designed in the first place (strict isolation, everything is in parallel) is perfectly fine for unit and integration testing. Just wanting to point out that this might feature-bloat FailGood here and it begins to lose its simplicity.

**Comment by christophsturm** [comment]
The idea here is not "sequential tests" but "tests with steps", maybe thats not well communicated in the api.
what is now called "describeSteps" is not a context, but a test, and all the its in it are a step.
Also the current api is not something that we can implement in a performant way, because we have to execute the "describeSteps" block at discovery time to get the names of the steps, and executing it at discovery time will make discovery slower.

so a better api would be an it block with a list of lambdas.
`it("can create a user and log it in", steps=listOf(step("create the user", { ...}), step("log in"), {})))` 

This would also better communicate that its not muliple sequential tests but one test with steps. (that show up in the test results)

@mervyn-mccreight about making it an extension: making the internal apis extendable is maybe more work and will make failgood more complex than just adding the feature. And even if it could become an extension at some point it would just start as an internal feature, and that would then drive the effort to make the internal apis extendable to move it out. 

TLDR: this pr will never be merged.
